File: benchmark/method_call.py
#!/usr/bin/python
# http://www.bagley.org/~doug/shootout/

import sys
import time

class Toggle(object):
    def __init__(self, start_state):
        self.bool = start_state
    def value(self):
        return(self.bool)
    def activate(self):
        self.bool = not self.bool
        return(self)

class NthToggle(Toggle):
    def __init__(self, start_state, max_counter):
        Toggle.__init__(self, start_state)
        self.count_max = max_counter
        self.counter = 0
    def activate(self):
        self.counter += 1
        if (self.counter >= self.count_max):
            super(NthToggle, self).activate()
            self.counter = 0
        return(self)


def main():
    start = time.clock()

    NUM = 1000000

    val = 1
    toggle = Toggle(val)
    for i in xrange(0,NUM):
        val = toggle.activate().value()
    if val:
        print "true"
    else:
        print "false"

    val = 1
    ntoggle = NthToggle(val, 3)
    for i in xrange(0,NUM):
        val = ntoggle.activate().value()
    if val:
        print "true"
    else:
        print "false"

    print("elapsed: " + str(time.clock() - start))


main()

File: benchmark/method_call.rb
#!/usr/bin/ruby
# -*- mode: ruby -*-
# $Id: methcall.ruby,v 1.1 2004-05-19 18:10:41 bfulgham Exp $
# http://www.bagley.org/~doug/shootout/
# with help from Aristarkh Zagorodnikov

class Toggle
    def initialize(start_state)
        @bool = start_state
    end

    def value
        @bool
    end

    def activate
        @bool = !@bool
        self
    end
end

class NthToggle < Toggle
    def initialize(start_state, max_counter)
        super start_state
        @count_max = max_counter
        @counter = 0
    end

    def activate
        @counter += 1
        if @counter >= @count_max
            super
            @counter = 0
        end
        self
    end
end

def main()
    start = Time.now

    n = 100000

    val = 1
    toggle = Toggle.new(val)
    n.times do
        val = toggle.activate().value()
        val = toggle.activate().value()
        val = toggle.activate().value()
        val = toggle.activate().value()
        val = toggle.activate().value()
        val = toggle.activate().value()
        val = toggle.activate().value()
        val = toggle.activate().value()
        val = toggle.activate().value()
        val = toggle.activate().value()
    end
    if val then puts "true" else puts "false" end

    val = 1
    ntoggle = NthToggle.new(val, 3)
    n.times do
        val = ntoggle.activate().value()
        val = ntoggle.activate().value()
        val = ntoggle.activate().value()
        val = ntoggle.activate().value()
        val = ntoggle.activate().value()
        val = ntoggle.activate().value()
        val = ntoggle.activate().value()
        val = ntoggle.activate().value()
        val = ntoggle.activate().value()
        val = ntoggle.activate().value()
    end
    if val then puts "true" else puts "false" end

    puts "elapsed: " + (Time.now - start).to_s
end

main()
